### PR TITLE
ARROW-17078: [C++] Clean up error handling in C++ Examples

### PR DESCRIPTION
**Pull request by vibhatha**
A simple util file was added to include a macro used in each example and may be in future we can include common utils like data generators, writers, etc in the here. This is just a suggestion.

### CODE REVIEW

**Comment by vibhatha** [comment]
cc @pitrou @lidavidm

**Comment by pitrou** [comment]
While usually I'm strongly in favour of avoiding duplication and building up reusable facilities, in this case I'm not sure it's a good idea to introduce a specific header file for examples, instead of a single `.cc` file per example.
@westonpace @cyb70289 @kou What do you think?

**Comment by lidavidm** [comment]
If anything I'd rather have examples ported to the cookbook, but yes, I'd rather try to keep examples here reasonably self-contained.

**Comment by kou** [comment]
In general, I want to keep examples as simple as possible for easy to understand by users. For example, minimal dependencies, straightforward implementations and so on.

In this case, I don't want to do this because it increases a (small) dependency (`example_utils.h`).

How about removing `ABORT_ON_FAILURE()` and using `ARROW_RETURN_NOT_OK()` with non-`main()` function like https://github.com/apache/arrow/blob/master/cpp/examples/arrow/join_example.cc instead of using `ABORT_ON_FAILURE()` in `main()`?

**Comment by vibhatha** [comment]
@kou may be I can do that here.

**Comment by kou** [comment]
Yes, please. :-)

**Comment by lidavidm** [comment]
We should also file a JIRA since this isn't really a minor change

**Comment by vibhatha** [comment]
@lidavidm I renamed the PR to match with the filed JIRA. Should we do something to track it here?

**Comment by lidavidm** [comment]
It will automatically link the JIRA, just be patient and let the Dev CI pipeline run to make the link.

File: cpp/examples/arrow/dataset_documentation_example.cc [hunk 1]
```diff
@@ -39,99 +39,96 @@ namespace ds = arrow::dataset;
...
-#define ABORT_ON_FAILURE(expr)                     \
-  do {                                             \
-    arrow::Status status_ = (expr);                \
-    if (!status_.ok()) {                           \
-      std::cerr << status_.message() << std::endl; \
-      abort();                                     \
-    }                                              \
-  } while (0);
+/**
+ * @brief Run Example
```
Comment (lidavidm) [inline cpp/examples/arrow/dataset_documentation_example.cc:43]: nit: use `\brief` for consistency with the C++ codebase

File: cpp/examples/arrow/dataset_documentation_example.cc [hunk 2]
```diff
@@ -194,10 +191,10 @@ std::shared_ptr<arrow::Table> FilterAndSelectDataset(
   auto dataset = factory->Finish().ValueOrDie();
   // Read specified columns with a row filter
   auto scan_builder = dataset->NewScan().ValueOrDie();
-  ABORT_ON_FAILURE(scan_builder->Project({"b"}));
-  ABORT_ON_FAILURE(scan_builder->Filter(cp::less(cp::field_ref("b"), cp::literal(4))));
+  ARROW_RETURN_NOT_OK(scan_builder->Project({"b"}));
+  ARROW_RETURN_NOT_OK(scan_builder->Filter(cp::less(cp::field_ref("b"), cp::literal(4))));
   auto scanner = scan_builder->Finish().ValueOrDie();
```
Comment (lidavidm) [inline cpp/examples/arrow/dataset_documentation_example.cc:196]: Can we also eliminate all uses of ValueOrDie?

Comment (lidavidm) [reply cpp/examples/arrow/dataset_documentation_example.cc:196]: That applies to other files here too.

File: cpp/examples/arrow/flight_grpc_example.cc [hunk 3]
```diff
@@ -84,20 +75,12 @@ class HelloWorldServiceImpl : public HelloWorldService::Service {
...
-int main(int argc, char** argv) {
-  gflags::ParseCommandLineFlags(&argc, &argv, true);
-
-  if (FLAGS_port < 0) {
-    // For CI
-    std::cout << "Must specify a port with -port" << std::endl;
-    return EXIT_SUCCESS;
-  }
-
+arrow::Status RunFlightGRPC() {
```
Comment (lidavidm) [inline cpp/examples/arrow/flight_grpc_example.cc:78]:
```suggestion
arrow::Status RunFlightGrpc() {
```

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
Looks like failures are primarily due to the Protobuf linkage issue.

**Comment by vibhatha** [comment]
> Looks like failures are primarily due to the Protobuf linkage issue.

Is this independent of this change?

**Pull request merged**